### PR TITLE
Fix k8s api retry logic

### LINKED ISSUE #627: A lot of exceptions in Namerd with k8s storage
Hi,

I have deployed `Namerd` into Kubernetes with such configs:
```yaml
apiVersion: v1
kind: Service
metadata:
  name: namerd-sync
spec:
  selector:
    app: namerd
  ports:
  - name: sync
    port: 4100
---
apiVersion: v1
kind: Service
metadata:
  name: namerd-api
spec:
  selector:
    app: namerd
  ports:
  - name: api
    port: 4180
---
metadata:
  name: d-tab.l5d.io # the hyphen is required by the Kubernetes API. This will be converted to the CamelCase name "DTab".
apiVersion: extensions/v1beta1
kind: ThirdPartyResource
description: stores dtabs used by Buoyant's `namerd` service
versions:
  - name: v1alpha1 # Do not change this value as it hardcoded in Namerd and doesn't work with other value.
---
kind: ConfigMap
apiVersion: v1
metadata:
  name: namerd-config
data:
  config.yml: |-
    admin:
      port: 9991
    storage:
      kind: io.l5d.k8s
      experimental: true
    namers:
      - kind: io.l5d.k8s
        experimental: true
        host: 127.0.0.1
        port: 8001
    interfaces:
      - kind: io.l5d.thriftNameInterpreter
        ip: 0.0.0.0
        port: 4100
      - kind: io.l5d.httpController
        ip: 0.0.0.0
        port: 4180
---
kind: ReplicationController
apiVersion: v1
metadata:
  name: namerd
spec:
  replicas: 1
  selector:
    app: namerd
  template:
    metadata:
      labels:
        app: namerd
    spec:
      dnsPolicy: ClusterFirst
      volumes:
        - name: namerd-config
          configMap:
            name: namerd-config
      containers:
        - name: namerd
          image: buoyantio/namerd:0.7.4
          args:
            - /io.buoyant/namerd/config/config.yml
            - -com.twitter.finagle.tracing.debugTrace=true
            - -log.level=DEBUG
          imagePullPolicy: Always
          ports:
            - name: sync
              containerPort: 4100
            - name: api
              containerPort: 4180
          volumeMounts:
            - name: "namerd-config"
              mountPath: "/io.buoyant/namerd/config"
              readOnly: true
        - name: kubectl
          image: buoyantio/kubectl:1.2.3
          args:
          - "proxy"
          - "-p"
          - "8001"
          imagePullPolicy: Always
```

When I try to get all available dtabs (which is empty after deployment) I start see a lot of errors complaining to not accesible api and some exceptions. And these errors still appear all the time wheather dtab(s) created or deleted.

Errors complain about not accessible `/apis/l5d.io/v1alpha1/namespaces/default/dtabs` but it accessible and I can get results from k8s. Looks strange.

Attaching complete log file...
[namerd-exceptions.txt](https://github.com/BuoyantIO/linkerd/files/456278/namerd-exceptions.txt)

### PR DESCRIPTION
**Pull request by adleong**
Fixes #627

# Problem

When a watch on a k8s resource is interrupted, the infinite retry filter will retry the interrupted future indefinitely.  

# Solution

Don't retry on interruptions.

### CODE REVIEW

File: k8s/src/main/scala/io/buoyant/k8s/resources.scala [hunk 1]
```diff
@@ -14,6 +14,15 @@ import com.twitter.util.{Closable, _}
...
   def close(deadline: Time) = client.close(deadline)
+
+  protected def parseResponse[O <: KubeObject: Manifest](rsp: http.Response): Future[O] = {
+    rsp.status match {
+      case http.Status.Successful(_) => Api.parse[O](rsp)
+      case http.Status.NotFound => Future.exception(Api.NotFound(rsp))
+      case http.Status.Conflict => Future.exception(Api.Conflict(rsp))
+      case _ => Future.exception(Api.UnexpectedResponse(rsp))
+    }
+  }
```
Comment (olix0r) [inline k8s/src/main/scala/io/buoyant/k8s/resources.scala]: Is there a reason that this logic shouldn't just be in Api.parse?

Comment (gtcampbell) [inline k8s/src/main/scala/io/buoyant/k8s/resources.scala]: That sounds reasonable to me as well.

Comment (adleong) [inline k8s/src/main/scala/io/buoyant/k8s/resources.scala]: 👍

File: k8s/src/main/scala/io/buoyant/k8s/Api.scala [hunk 2]
```diff
@@ -56,8 +56,14 @@ object Api {
...
-    getContent(rsp).flatMap { content =>
-      Future.const(Json.read[T](content))
+    rsp.status match {
+      case http.Status.Successful(_) =>
+        getContent(rsp).flatMap { content =>
+          Future.const(Json.read[T](content))
+        }
+      case http.Status.NotFound => Future.exception(Api.NotFound(rsp))
+      case http.Status.Conflict => Future.exception(Api.Conflict(rsp))
+      case _ => Future.exception(Api.UnexpectedResponse(rsp))
```
Comment (olix0r) [inline k8s/src/main/scala/io/buoyant/k8s/Api.scala]: tioli: `Api.` is now redundant since this is within `Api`

**Pull request merged**